src/replication.c

```diff
@@ -0,0 +1,1067 @@
+	}
+}
+int
+handle_read_resp(spec_t *spec, replica_t *replica, zvol_io_hdr_t *io_rsp, void *data) { 
+	int luworker_id;
+	bool io_found = false;
+	rcmd_t *rep_cmd = NULL;
+	rcommon_cmd_t *rcommq_ptr = NULL;
+	MTX_LOCK(&replica->r_mtx);
+	//Find IO in read queue, signal luworker, and dequeue
+	TAILQ_FOREACH(rep_cmd, &replica->read_waitq, rread_cmd_next) {
+		if(io_rsp->io_seq == rep_cmd->io_seq) {
+			io_found = true;
+			rep_cmd->ack_recvd = true;
+			TAILQ_REMOVE(&replica->read_waitq, rep_cmd, rread_cmd_next);
+			MTX_UNLOCK(&replica->r_mtx);
+			rcommq_ptr = rep_cmd->rcommq_ptr;
+			rcommq_ptr->status = 1;
+			rcommq_ptr->data = data;
+			rcommq_ptr->data_len = io_rsp->len;
+			luworker_id = rcommq_ptr->luworker_id;
+			free(rep_cmd);
+
+			MTX_LOCK(&spec->luworker_rmutex[luworker_id]);
+			pthread_cond_signal(&spec->luworker_rcond[luworker_id]);
+			MTX_UNLOCK(&spec->luworker_rmutex[luworker_id]);
+
+			MTX_LOCK(&spec->rcommonq_mtx);
+			TAILQ_REMOVE(&spec->rcommon_waitq, rcommq_ptr, wait_cmd_next);
+			if(rcommq_ptr->completed == 1) {
+				free(rcommq_ptr);
+			} else {
+				rcommq_ptr->completed = 1;
+			}
+			MTX_UNLOCK(&spec->rcommonq_mtx);
+			MTX_LOCK(&replica->r_mtx);
+			break;
+		}
+	}
+	MTX_UNLOCK(&replica->r_mtx);
+	if(!io_found) {
+		return -1;
+	}
+	return 0;
+}
+
+int
+handle_write_resp(spec_t *spec, replica_t *replica, zvol_io_hdr_t *io_rsp) {
+	int luworker_id, i;
+	rcmd_t *rep_cmd = NULL;
+	rcommon_cmd_t *rcommq_ptr;
+	MTX_LOCK(&replica->r_mtx);
+	TAILQ_FOREACH(rep_cmd, &replica->waitq, rwait_cmd_next) {
+		if(rep_cmd->io_seq == io_rsp->io_seq) {
+			break;
+		}
+	}
+	if(rep_cmd == NULL) { 
+		REPLICA_ERRLOG("rep_cmd not found io_seq:%lu\n", io_rsp->io_seq);
+		MTX_UNLOCK(&replica->r_mtx);
+		return -1;
+	}
+	rep_cmd->ack_recvd = true;
+	if(rep_cmd->wrio_seq == replica->least_recvd + 1) {
+		while((rep_cmd = TAILQ_FIRST(&replica->waitq))) {
+			rcommq_ptr = rep_cmd->rcommq_ptr;
+			if(rep_cmd->ack_recvd == true) {
+				replica->least_recvd++;
+				MTX_UNLOCK(&replica->r_mtx);
+				rcommq_ptr->acks_recvd++;
+				luworker_id = rcommq_ptr->luworker_id;
+
+				rcommq_ptr->bitset &= ~(1 << replica->id);
+				if (rcommq_ptr->acks_recvd == spec->consistency_count) {
+					rcommq_ptr->status = 1;
+					//Transfer cmd from waitq to pendingq and wakeup luworker
+					MTX_LOCK(&spec->rcommonq_mtx);
+					TAILQ_REMOVE(&spec->rcommon_waitq, rcommq_ptr, wait_cmd_next);
+					if (rcommq_ptr->acks_recvd != spec->replica_count) {
+						TAILQ_INSERT_TAIL(&spec->rcommon_pendingq, rcommq_ptr, pending_cmd_next);
+					} else {
+						for (i=1; i < rcommq_ptr->iovcnt + 1; i++) {
+							xfree(rcommq_ptr->iov[i].iov_base);
+						}
+						rcommq_ptr->completed = 1;
+					}
+					MTX_UNLOCK(&spec->rcommonq_mtx);
+
+					MTX_LOCK(&spec->luworker_rmutex[luworker_id]);
+					pthread_cond_signal(&spec->luworker_rcond[luworker_id]);
+					MTX_UNLOCK(&spec->luworker_rmutex[luworker_id]);
+
+				} else if (rcommq_ptr->acks_recvd == spec->replica_count) {
```

Comment:
maintain the count to which IOs are sent, rather than looking at replica_count which might change